### ISSUE TITLE
proposal: build in hot-reload (templ generate --watch)

### ISSUE THREAD

**Issue by a-h** [a-h/templ#82]
The documentation suggests using [air](https://github.com/cosmtrek/air) - https://templ.guide/commands-and-tools/hot-reload

But it's installing another tool, and having to configure it. Running `templ generate --watch` would be a better user experience.

It would make #80 irrelevant, since it would be able to maintain internal state such as the last-updated time. https://github.com/fsnotify/fsnotify seems to be the library to use.

Since templ doesn't know how the web server is configured, it can't easily inject client side code to force a refresh.

But... templ could ship with HTTP middleware that implements hot reload. The middleware would be able to set the target directory that contains the templates and use fsnotify to watch it, or figure it the target directory from `os.Args`, run `templ generate` and then start a replacement of itself based on the `os.Args`. Since it's HTTP middleware, it could parse the output HTML and inject stuff into the page.

Alternatively, it could break the problem into two parts... so that `templ generate --watch` starts a web server on localhost which controls hot reload, and it also ships a middleware component that injects a script to connect to the `templ generate --watch` port (e.g. http://localhost:14734/) and deals with CORS.

**Comment by gedw99** [comment]
https://github.com/loov/watchrun might be a good one to do this.

**Comment by gedw99** [comment]
> The documentation suggests using [air](https://github.com/cosmtrek/air) - https://templ.guide/commands-and-tools/hot-reload
> 
> But it's installing another tool, and having to configure it. Running `templ generate --watch` would be a better user experience.
> 
> It would make #80 irrelevant, since it would be able to maintain internal state such as the last-updated time. https://github.com/fsnotify/fsnotify seems to be the library to use.
> 
> Since templ doesn't know how the web server is configured, it can't easily inject client side code to force a refresh.
> 
> But... templ could ship with HTTP middleware that implements hot reload. The middleware would be able to set the target directory that contains the templates and use fsnotify to watch it, or figure it the target directory from `os.Args`, run `templ generate` and then start a replacement of itself based on the `os.Args`. Since it's HTTP middleware, it could parse the output HTML and inject stuff into the page.
> 
> Alternatively, it could break the problem into two parts... so that `templ generate --watch` starts a web server on localhost which controls hot reload, and it also ships a middleware component that injects a script to connect to the `templ generate --watch` port (e.g. http://localhost:14734/) and deals with CORS.

Like the thinking here - clean and flexible.

**Comment by delaneyj** [comment]
I use https://github.com/go-rod/rod and a taskfile watch

**Comment by joerdav** [comment]
@delaneyj  I think the main proposal here is to add some sort of middleware that not only will rebuild templates but will also refresh the browser during development. Similar to this proxy: https://github.com/diamondburned/saq

**Comment by gedw99** [comment]
> @delaneyj I think the main proposal here is to add some sort of middleware that not only will rebuild templates but will also refresh the browser during development. Similar to this proxy: https://github.com/diamondburned/saq

saq only works on Linux due to its dependency on [illarion/gonotify](https://github.com/illarion/gonotify).

**Comment by delaneyj** [comment]
> @delaneyj I think the main proposal here is to add some sort of middleware that not only will rebuild templates but will also refresh the browser during development. Similar to this proxy: https://github.com/diamondburned/saq

Yeah that's what's I'm doing, the combination of `task -w` and `rod` is how I'm doing hot reload

**Comment by joerdav** [comment]
@delaneyj Oh nice!

**Comment by delaneyj** [comment]
@joerdav 

```go

func runHotReload(port int, onStartPath string) toolbelt.CtxErrFunc {
	return func(ctx context.Context) error {
		onStartPath = strings.TrimPrefix(onStartPath, "/")
		localHost := fmt.Sprintf("http://localhost:%d", port)
		localURLToLoad := fmt.Sprintf("%s/%s", localHost, onStartPath)

		// Make sure page is ready before we start
		backoff := backoff.NewExponentialBackOff()
		for {
			if _, err := http.Get(localURLToLoad); err == nil {
				break
			}

			d := backoff.NextBackOff()
			log.Printf("Server not ready. Retrying in %v", d)
			time.Sleep(d)
		}

		// Launch browser in user mode, so we can reuse the same browser session
		wsURL := launcher.NewUserMode().MustLaunch()
		browser := rod.New().ControlURL(wsURL).MustConnect().NoDefaultDevice()

		// Get the current pages
		pages, err := browser.Pages()
		if err != nil {
			return fmt.Errorf("failed to get pages: %w", err)
		}
		var page *rod.Page
		for _, p := range pages {
			info, err := p.Info()
			if err != nil {
				return fmt.Errorf("failed to get page info: %w", err)
			}

			// If we already have the page open, just reload it
			if strings.HasPrefix(info.URL, localHost) {
				p.MustActivate().MustReload()
				page = p
				break
			}
		}
		if page == nil {
			// Otherwise, open a new page
			page = browser.MustPage(localURLToLoad)
		}

		// time.Sleep(1 * time.Second)
		// page.Reload()
		// page.Reload()
		// page.Reload()

		slog.Info("page loaded", "url", localURLToLoad, "page", page.TargetID)
		return nil
	}
}
```

And within the taskfile
```yml
  racer_dev:
    desc: Server hot reload
    sources:
      - "**/*.go"
    deps:
      - racer_tw
    cmds:
      - killall -q foo_bin  || echo "Process was not running."
      - go mod tidy
      - go build -o ./foo_bin cmd/foo/*.go
      - ./foo_bin
    ```

**Comment by a-h** [comment]
I've had chance to have a crack at this.

The `fsnotify` implementation didn't produce all events on my Mac. From reading around there's a long outstanding issue with it around needing a backup process and my understanding is that some frameworks don't rely on it, and use a 250ms delay between `os.Walk` instead. I've implemented that, and it seems to work well. I've put in some basic ignore rules (node_modules, .git etc.) to make it skip directories that should never contain templ files.

Unfortunately, the `rod` implementation produces panics on my Mac. Even, if I closed all the browsers and let it try again. Maybe it's because I have multiple Chrome profiles, but it didn't seem reliable to me for the use case, so I've attempted the proxy approach and used the `browser` package to auto-open the proxy endpoint in the browser.

I have a working version in the https://github.com/a-h/templ/tree/hot_reload branch.

I can run `templ generate --watch --proxy="http://localhost:8080" --cmd="runtest"` and it will start a proxy on localhost through to `localhost:8080`, and, provided any response has a `Content-Type` of `text/html` and the HTML contains a `</body>` tag, the proxy will inject an automatic reload script.

The reload script connects to the proxy's SSE endpoint. Any changes to `templ` files results in a recompilation, this recompilation then sends a message to the SSE handler, and any connected clients then execute `window.location.reload()`.

I haven't had time to tidy up the code much, but it works on a basic test web server (`runtest`).

I've made a PR - feel free to comment: https://github.com/a-h/templ/pull/130
